lib/configurators/aws/aws.go

```diff
@@ -70,39 +71,71 @@ type databaseActions struct {
 	// iamAuth is a list of actions used for enabling IAM auth.
 	iamAuth []string
 	// metadata is a list of actions used for fetching database metadata
-	// (excluding the ones already in "discovery").
 	metadata []string
 	// managedUsers is a list of actions used for managing database users.
 	managedUsers []string
-	// boundary is a list of actions only used for boundary policies.
-	boundary []string
+	// authBoundary is a list of actions for authorization that need to added
+	// to boundary policies.
+	authBoundary []string
 
 	requireIAMEdit        bool
 	requireSecretsManager bool
 }
 
-func (a databaseActions) buildStatementForDiscovery() *awslib.Statement {
-	// Note that currently extra boundary policies are not required for discovery service.
+func (a databaseActions) buildStatement(opt databaseActionsBuildOption) *awslib.Statement {
+	var actions []string
+	if opt.withDiscovery {
+		actions = append(actions, a.discovery...)
+	}
+	if opt.withMetadata {
+		actions = append(actions, a.metadata...)
+	}
+	if opt.withAuth {
+		actions = append(actions, a.iamAuth...)
+		actions = append(actions, a.managedUsers...)
+	}
+	if opt.withAuthBoundary {
+		actions = append(actions, a.authBoundary...)
+	}
 	return &awslib.Statement{
 		Effect:    awslib.EffectAllow,
-		Actions:   a.discovery,
+		Actions:   apiutils.Deduplicate(actions),
 		Resources: []string{"*"},
 	}
 }
 
-func (a databaseActions) buildStatement(boundary bool) *awslib.Statement {
-	var actions []string
-	actions = append(actions, a.discovery...)
-	actions = append(actions, a.iamAuth...)
-	actions = append(actions, a.metadata...)
-	actions = append(actions, a.managedUsers...)
-	if boundary {
-		actions = append(actions, a.boundary...)
-	}
-	return &awslib.Statement{
-		Effect:    awslib.EffectAllow,
-		Actions:   actions,
-		Resources: []string{"*"},
+type databaseActionsBuildOption struct {
+	withDiscovery    bool
+	withMetadata     bool
+	withAuth         bool
+	withAuthBoundary bool
+}
+
+func makeDatabaseActionsBuildOption(flags configurators.BootstrapFlags, targetCfg targetConfig, boundary bool) databaseActionsBuildOption {
```

Comment:
Is it possible for this to be called with both `flags.DiscoveryService` and `flags.DiscoveryServiceConfig` set to true, and if so is there a particular reason we want `flags.DiscoveryService` to take precedence?

It looks like both flags get set to `true` here:

https://github.com/gravitational/teleport/blob/2b39df8fd1853ed314c61d7966827afa87c0e203/tool/teleport/common/teleport.go#L520-L523

Reply:
Changed to enum as discussed below.